apps/rush-lib/src/logic/buildCache/ProjectBuildCache.ts

```diff
@@ -134,30 +153,46 @@ export class ProjectBuildCache {
       )
     );
 
-    const tarStream: stream.Writable = tar.extract({ cwd: projectFolderPath });
-    const extractTarPromise: Promise<boolean> = new Promise(
-      (resolve: (result: boolean) => void, reject: (error: Error) => void) => {
-        try {
-          tarStream.on('error', (error: Error) => reject(error));
-          tarStream.on('close', () => resolve(true));
-          tarStream.on('drain', () => resolve(true));
-          tarStream.write(cacheEntryBuffer);
-        } catch (e) {
-          reject(e);
-        }
+    const tarExecutablePath: string | undefined = ProjectBuildCache._tarExecutablePath;
+    let restoreSuccess: boolean;
+    if (!tarExecutablePath || !localCacheEntryPath) {
+      if (!cacheEntryBuffer && localCacheEntryPath) {
+        // eslint-disable-next-line require-atomic-updates
+        cacheEntryBuffer = await FileSystem.readFileToBufferAsync(localCacheEntryPath);
       }
-    );
 
-    let restoreSuccess: boolean;
-    let updateLocalCacheSuccess: boolean;
-    if (setLocalCacheEntryPromise) {
-      [restoreSuccess, updateLocalCacheSuccess] = await Promise.all([
-        extractTarPromise,
-        setLocalCacheEntryPromise
-      ]);
+      if (!cacheEntryBuffer) {
+        throw new Error('Expected the cache entry buffer to be set.');
+      }
+
+      // If we don't have tar on the PATH, or if we failed to update the local cache entry,
+      // untar in-memory
+      const tarStream: stream.Writable = tar.extract({ cwd: projectFolderPath });
+      restoreSuccess = await new Promise(
+        (resolve: (result: boolean) => void, reject: (error: Error) => void) => {
+          try {
+            tarStream.on('error', (error: Error) => reject(error));
+            tarStream.on('close', () => resolve(true));
```

Comment:
FYI, you can simplify this pattern as:
```ts
import { once } from 'events'; // Built-in

const tarPromise = once('drain', tarStream); // This automatically binds "error" to the promise rejection as well
tarStream.write(cacheEntryBuffer);
await tarPromise;

Reply:
https://nodejs.org/docs/latest-v10.x/api/events.html#events_events_once_emitter_name